### PR TITLE
Format messages correctly and support /me commands

### PR DESCRIPTION
**Pull request by jvperrin**
Unfortunately this contains changes made for #44, but a small part of this pull request depends on those changes, so I've based it off of that branch. For now, I'll make this pull request for merging into that branch, so the changes shown reflect more accurately what is actually changing in the single commit relevant to this pull request, but later on this should be merged into master after #44 has been merged.

A Christmas gift for both IRC and Slack users! 🎁 🎄 🎁

### CODE REVIEW

File: slackbridge/bots.py [hunk 1]
```diff
@@ -71,8 +70,12 @@ def resolve(self):
                 channel_name = IRCBot.channels[channel_id]['name']
                 if message_type == 'message':
                     if 'subtype' in self.raw_message:
-                        if self.raw_message['subtype'] in IGNORED_MSG_SUBTYPES:
+                        subtype = self.raw_message['subtype']
+                        if subtype in IGNORED_MSG_SUBTYPES:
```
Comment (kkuehler) [inline slackbridge/bots.py]: Hmm. I wonder if we should do this check *before* we put messages in the queue.

File: slackbridge/bots.py [hunk 2]
```diff
@@ -150,24 +163,27 @@ def signedOn(self):
...
+
+    def action(self, user, channel, message):
+        self.post_to_slack(user, channel, '_{}_'.format(message))
+
+    def post_to_slack(self, user, channel, message):
+        # User is like 'jvperrin!Jason@fireball.ocf.berkeley.edu' so only
         # take the part before the exclamation mark for the Slack display name
         assert user.count('!') == 1
-        user_nick, _ = user.split('!')
+        user, _ = user.split('!')
```
Comment (kkuehler) [inline slackbridge/bots.py]: Minor, but I feel that this should be a method in utils.

File: slackbridge/utils.py [hunk 3]
```diff
@@ -36,6 +74,54 @@ def strip_nick(nick):
     ])
 
 
+def format_irc_message(text, users, channels):
```
Comment (kkuehler) [inline slackbridge/utils.py]: very well done.

File: slackbridge/bots.py (hunk 1 above)
Comment (redplanks) [reply slackbridge/bots.py]: In either case, if we wanted to add handling for another type, we'd have to modify the list and add a handler. So it's largely a question of organization.

File: slackbridge/bots.py [hunk 4]
```diff
@@ -93,9 +96,19 @@ def _change_presence(self, user_bot):
             elif self.raw_message['presence'] == 'active':
                 user_bot.back()
 
+        def _irc_action(self, channel_name, user_bot):
+            user_bot.post_to_irc(
+                user_bot.describe,
```
Comment (redplanks) [inline slackbridge/bots.py]: Where does "describe" come from? I can't find any references to it.

Comment (jvperrin) [reply slackbridge/bots.py]: http://twistedmatrix.com/documents/current/api/twisted.words.protocols.irc.IRCClient.html#describe

It seems like the most appropriate way to do `/me` actions anyway.

File: slackbridge/bots.py [hunk 5]
```diff
@@ -93,9 +96,19 @@ def _change_presence(self, user_bot):
             elif self.raw_message['presence'] == 'active':
                 user_bot.back()
 
+        def _irc_action(self, channel_name, user_bot):
```
Comment (redplanks) [inline slackbridge/bots.py]: The method title seems to give an impression that it's a base method to extend off of. Maybe make it a bit more explicit with "_irc_me_action" or something along those lines?

File: slackbridge/bots.py (hunk 2 above)
Comment (jvperrin) [reply slackbridge/bots.py]: Yeah, that would make sense and also make this easier to actually run tests on. I think if we can move more stuff into utils or other helper methods then that will be good overall.

File: slackbridge/bots.py (hunk 1 above)
Comment (jvperrin) [reply slackbridge/bots.py]: Yeah, we could just not add it to the queue, but we have other message types we skip in the same way after we add them, so I think if we were going the route of not adding them to the queue we should do it to those messages too.

File: slackbridge/utils.py [hunk 6]
```diff
@@ -36,6 +74,54 @@ def strip_nick(nick):
...
+    text = text.replace('&lt;', '<').replace('&gt;', '>').replace('&amp;', '&')
+    return text
+
+
+def format_slack_message(text):
+    # Strip any color codes coming from IRC, since Slack cannot display them
+    # Current solution is taken from https://stackoverflow.com/a/970723
+    # TODO: Preserve bold and italics (convert to markdown?)
+    return re.sub(r'\x03(?:\d{1,2}(?:,\d{1,2})?)?', '', text, flags=re.UNICODE)
+
```
Comment (redplanks) [inline slackbridge/utils.py]: I think Slack also requires that &, <, and > be [HTML-encoded](https://api.slack.com/docs/message-formatting).

File: slackbridge/utils.py (hunk 3 above)
Comment (jvperrin) [reply slackbridge/utils.py]: Thanks! It's mostly adapted from the [slack-irc project](https://github.com/ekmartin/slack-irc/blob/2b5ceb7ca7beb/lib/bot.js#L154) that has to accomplish a similar task.

File: slackbridge/bots.py (hunk 5 above)
Comment (jvperrin) [reply slackbridge/bots.py]: Good point, that would be clearer, I'll change it.

File: slackbridge/utils.py (hunk 6 above)
Comment (jvperrin) [reply slackbridge/utils.py]: Their docs do say that, but I believe the client library we are using to interface with Slack already does this.

Comment (redplanks) [reply slackbridge/utils.py]: Ah, cool.

Comment (jvperrin) [reply slackbridge/utils.py]: I can't find exactly where it does, but we haven't had any issues with sending these characters to Slack, and I tested them yesterday too.

Comment (jvperrin) [reply slackbridge/utils.py]: Ok, I've added a comment about this, since I think it's a good thing to mention.

File: slackbridge/bots.py (hunk 2 above)
Comment (jvperrin) [reply slackbridge/bots.py]: Alright, I split this off into a separate utils function.

**Pull request merged**